### DISCUSSION TITLE
Thin clients

### DISCUSSION THREAD

**Discussion by llantones** [ltsp/ltsp#243] [Q&A]
Hello:
If I want to use thin client that run the applications in server I must use ltsp5?
Thanks

**Comment by alkisg** [comment]
In the very long past, Xorg was "network transparent" and drawing things on the screen involved functions like "draw an arc, a rectangle, a font etc". I.e. the required LAN bandwidth was minimal.

Then multimedia and compositing and OpenGL and Wayland etc came, and Xorg developers stopped caring about network transparency, and things stopped working well over the network.

For example, an HD video 1920x1080x32bpp@60fps needs 3981312000 bps, i.e. 4 gigabits per second of LAN bandwidth. If 10 clients need to watch a full HD video without dropped frames, you'd need a 40 gigabit network card on the LTSP server.

So developers stopped caring about thin clients and focused on "remote desktop" for the cases where this is needed. For example, RemoteFX in Windows RDP would autodetect the video stream and send it to the client in compressed form (e.g. vp9), thus needing e.g. 1 mbps instead of 4 gbps. And RDP also supports sharing folders both ways, and local/remote printing etc, everything that LTSP clients supported.

To sum up:

With the new LTSP, one can netboot clients, and then use "remote desktop" to run applications on the server: xfreerdp, x2go etc. [This comment](https://github.com/ltsp/ltsp/issues/39#issuecomment-539150708) gives a quick how-to. It should also be possible to do remote Xorg with the "X -query" command; but ltspfs, localapps, remoteapps etc won't be implemented in the new LTSP.

With the old LTSP one can still do remote Xorg, ltspfs, localapps, remoteapps; but no developer works on those anymore, so users will need to resolve any problems that appear.

**Comment by llantones** [comment]
Very clear explanation, thanks. 
I want to use only one client, in this case, if I use a slow pentium 4 is better ltsp5?
Thanks.

**Comment by alkisg** [comment]
To give better advice, please mention your server/client/network/use case [specs](https://github.com/ltsp/ltsp/wiki/specs).

**Comment by llantones** [comment]
Ok, sorry:
 - 1 server: [Ryzen 5 1600x with 13196 cpubenchmark score ](https://www.cpubenchmark.net/cpu.php?cpu=AMD+Ryzen+5+1600X&id=3000) and 8 GB RAM
- 1 client: [Pentium 4 with 164 cpubenchmark score](https://www.cpubenchmark.net/cpu.php?cpu=Intel+Pentium+4+1.40GHz&id=1688) and 512 KB RAM
- 1 Ethernet Crossover Cable T568A-T568B
- Software: LTSP version 19.09, distribution Ubuntu 18.04.3, flavour Lubuntu, kernel 4.15
- Use case: running libreoffice and browsing.

**Comment by alkisg** [comment]
You have very powerful server and very weak client (if it's actually 1.4 GHz), so it's best to use it as a thin client or with remote desktop.
So either LTSP5 in thin mode (might be a tiny bit better), or LTSP19 and xfreerdp/x2go.

If you had a Pentium 4 at 2.4 GHz and 1 GB RAM, then it'd be better to use the normal LTSP19 "fat" client mode.

**Comment by llantones** [comment]
OK

**Comment by alkisg** [comment]
The http://packages.ubuntu.com/lightdm-remote-session-freerdp2 package, if it works as advertised, might allow a user to login to a thin session without any other configuration. Just by installing xrdp on the LTSP server.

**Comment by llantones** [comment]
Hello alkisg:
Now I use Debian 10 and ltsp5:
https://wiki.debian.org/LTSP/Howto
and all works perfect and very fast. 

I want ltsp to use old pc and only need libreoffice and internet through firefox and thundenbird. I don't have the problem of the videos. Therefore I  am interesting in thin client, I like in the future can use thin client. I will test lightdm-remote-session-freerdp2 in ltsp19 and write here the results.
Thanks

**Comment by llantones** [comment]
Hello:
I installed from scratch lubuntu 18.10., then installed ltsp19:
https://ltsp.github.io/docs/installation/
Then I probed boot the client and all ok.
Then I installed http://packages.ubuntu.com/lightdm-remote-session-freerdp2
but I don't see how to select the thin session. Where is the information about this?
Thanks

**Comment by alkisg** [comment]
The basic idea for thin clients is:

- Install xrdp on the server
- Run `xfreerdp <params> /f /v:server` from the client

xfreerdp is just a single command, one can run it via multiple ways. For example others here do it with an LTSP guest account, who autologins and automatically launches just xfreerdp:
https://github.com/ltsp/community/wiki/xfreerdp
https://github.com/ltsp/community/wiki/RDP-Kiosks-on-Ubuntu

The persons at the [arctica project](https://github.com/ArcticaProject) have developed that greeter above ^ which is also supposed to run xfreerdp. I haven't used it; if you choose this path you would need to either read their documentation or contact them for support.

Personally I do not have any thin client setups (and Ubuntu stopped the 32bit architecture anyway) so I do not have any motivation to document or support such use cases. But of course others are very welcomed to contribute documentation or code or even join the development team and support thin clients out of the box.

**Comment by llantones** [comment]
> Ubuntu stopped the 32bit architecture anyway

I don't understand where is it neccesary 32 bit? I thought it is used for 486 but  Pentium 4 can use 64 bits and I reutilize Celeron and newer.

**Comment by alkisg** [comment]
Approximately,

- Pentium 4 up to 2.4 GHz are 32 bit and run better as thin clients.
- Pentium 4 from 2.4 GHz to 3 GHz are 32 bit and run better as fat clients.
- Pentium 4 from 3 GHz and up may be 64 bit and run better as fat clients.

So, if your clients are 64 bit, why would you run them as thin clients? They would surely run better as diskless, netbooted LTSP fat clients, which is the default for the new LTSP.

I can understand the RDP use case (e.g. need for Windows desktop), but not the "remote Xorg" for 64 bit fast clients that work better as fat...

**Comment by quiescer** [comment]
One reason we use thin clients is to share the RAM of the ltsp server. Our
terminals may have only 1 GB RAM (possibly even less in some cases).

**Comment by alkisg** [comment]
Please read https://github.com/ltsp/community/issues/32#issuecomment-541459584.

I think if anyone has 64bit clients, it'd be most wise to get a bit of RAM for them (if they don't already have 2+ GB as most 64bit clients), rather than invest in 10+ gigabit networking.

But don't get me wrong, the xfreerdp / x2go implementations still work fine, so you can still have thin clients, see the wiki pages links above.
It's the "remote Xorg with port forwarding" implementation that's not supported. Well, if you insist, you should  even be able to use https://wiki.archlinux.org/index.php/XDMCP, but without the LTSP5 ltspfs/localapps/remoteapps as no developer cared to maintain them.

Finally, if some developer wants to join the upstream team and support those, or if funding is found, then sure, those can be supported too; but currently noone cares to do that.

**Comment by ajparag** [comment]
Is it possible to connect ncomputing L230 thin client with LTSP?

**Comment by alkisg** [comment]
To be able to run any version of LTSP, clients must be able to run Linux.
AFAIK this client you mentioned is a "zero client" which isn't able to run Linux.

**Comment by ajparag** [comment]
thanks a lot for your quick response. I am a complete noob. I really appreciate your comment. Thanks again.

**Comment by maresb** [comment]
I have not tried it out yet, but X2Go offers two versions of a "Thin Client Environment" ([one version](https://wiki.x2go.org/doku.php/wiki:advanced:tce:install) and [another version](https://wiki.x2go.org/doku.php/doc:howto:tce)) which include a PXE Server which serves bootable X2Go thin client images.  If you only have a few clients and don't mind clogging your LAN with RDP traffic, X2Go TCE is probably much simpler and better suited than LTSP.

Update: the setup procedure looked fairly complicated, so I decided it was easiest to just use a LiveUSB with X2Go installed.
